app/models/amazon_agent_manager.rb

```diff
@@ -0,0 +1,411 @@
+    role = @iam.create_role({
+      role_name: role_name,
+      assume_role_policy_document: policy_doc.to_json
+    })
+
+    # grant all priviledges
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonS3FullAccess'
+    })
+
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonEC2FullAccess'
+    })
+
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonSQSFullAccess'
+    })
+
+    role
+  end
+
+  def create_profile(profile_name = MIQ_SSA, role_name = MIQ_SSA)
+    ssa_profile = @iam.instance_profile(profile_name)
+    ssa_profile = @iam.create_instance_profile(instance_profile_name: profile_name) unless ssa_profile.exists?
+
+    find_or_create_role(role_name)
+    ssa_profile.add_role(role_name: role_name) if ssa_profile.roles.size == 0
+
+    ssa_profile
+  end
+
+  def get_agent_image_id
+    imgs = @ec2.client.describe_images(
+      filters: [
+        {
+          name: "name",
+          values: [ "RHEL-7.3_HVM_GA*" ]
+        }
+      ]
+    ).images
+
+    imgs[0].image_id
+  end
+
+  def create_security_group(group_name = MIQ_SSA)
+    begin
+      sgs = @ec2.client.describe_security_groups(
+        filters: [
+          {
+            name: "group-name",
+            values: [ group_name ]
+          }
+        ]
+      ).security_groups
+      return sgs[0].group_id if sgs.length > 0
+
+      # create security group if not exist
+      security_group = @ec2.create_security_group({
+        group_name: group_name,
+        description: 'Security group for MIQ SSA Agent',
+        vpc_id: @ec2.client.describe_vpcs.vpcs[0].vpc_id
+      })
+
+      security_group.authorize_ingress({
+        ip_permissions: [{
+          ip_protocol: 'tcp',
+          from_port: 22,
+          to_port: 22,
+          ip_ranges: [{
+            cidr_ip: '0.0.0.0/0'
+          }]}]
+      })
+
+      security_group.authorize_ingress({
+        ip_permissions: [{
+          ip_protocol: 'tcp',
+          from_port: 80,
+          to_port: 80,
+          ip_ranges: [{
+            cidr_ip: '0.0.0.0/0'
+          }]}]
+      })
+
+      security_group.authorize_ingress({
+        ip_permissions: [{
+          ip_protocol: 'tcp',
+          from_port: 443,
+          to_port: 443,
+          ip_ranges: [{
+            cidr_ip: '0.0.0.0/0'
+          }]}]
+      })
+
+      security_group.group_id
+    end
+  end
+
+  # Get Key Pair for SSH. Create a new one if not exists.
+  def get_key_pair(pair_name = MIQ_SSA)
+    kp = Authentication.where(name: pair_name)
+    return kp[0] if kp.length > 0 && kp[0].resource_id == ems.id
+
+    ManageIQ::Providers::CloudManager::AuthKeyPair.create_key_pair(ems.id,
+      { :key_name => pair_name })
+  end
+
+  def create_pem_file(pair_name = MIQ_SSA)
+    kp = get_key_pair(pair_name)
+    pem_file_name = pair_name+".pem"
+    File.open(pem_file_name, 'w') {|f| f.write(kp.auth_key) }
+    File.chmod(0400, pem_file_name)
+    pem_file_name
+  end
+
```

Comment:
I think all of these script generation routines should be removed and replaced with an actual shell script file.
The script file should probably reside in the amazon_ssa_support  gem. The agent manager should then
read the contents of the script file through a method call into that gem (probably a module method).

Remember, the script file will probably be different between upstream and downstream, so the inplementation will need to be able to account for this.